File: src/miro_drive.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <xycar_msgs/xycar_motor.h>
#include <time.h>
#include <vector>
#include <random>
#include <cmath>
#include <iostream>
#include <numeric>

using namespace std;

vector<float> Dist;
int avg_val[5]; //직선 ,fl, fr, l, r
int dic_ang[5] = { 0, -50, 50, -50, 50};
ros::Publisher pub;

//평균
void cal_avg(int (&ref)[5]){
    //front
    int fnum = 42;
    int fsum = 0;
    for(int i = -21; i < 21; i++){
        if(Dist[i] == 0){
            fnum--;
        }
        else{
            fsum += Dist[i];
        }
    }
    ref[0] = fsum/fnum;
    //front_left
    int flnum = 84;
    int flsum = 0;
    for(int i = -105; i < -21; i++){
        if(Dist[i] == 0){
            flnum--;
        }
        else{
            flsum += Dist[i];
        }
    }
    ref[1] = flsum/flnum;
    //front_right
    int frnum = 84;
    int frsum = 0;
    for(int i = 21; i < 105; i++){
        if(Dist[i] == 0){
            frnum--;
        }
        else{
            frsum += Dist[i];
        }
    }
    ref[2] = frsum/frnum;
    //left
    int lnum = 42;
    int lsum = 0;
    for(int i = -147; i < -105; i++){
        if(Dist[i] == 0){
            lnum--;
        }
        else{
            lsum += Dist[i];
        }
    }
    ref[3] = lsum/lnum;
    //right
    int rnum = 42;
    int rsum = 0;
    for(int i = 105; i < 147; i++){
        if(Dist[i] == 0){
            rnum--;
        }
        else{
            rsum += Dist[i];
        }
    }
    ref[4] = rsum/rnum;
}



// 1.8 이내 1.8로 반환
/*
void change_val()
{
    for (int i = 0; i < Dist.size(); i++)
    {
        if (Dist[i] <= 1.8)
        {
            Dist[i] = 1.8;
        }
    }
    return;
}
*/

void callback(const sensor_msgs::LaserScan::ConstPtr& data)
{
    Dist = data->ranges; //.begin(), data->ranges.end());
    //std::cout<< Dist.size() <<"\n"
}

// motor_msg에 speed, angle 담아서 전송
void drive_go (int Speed, int Angle)
{
    xycar_msgs::xycar_motor msg;

    msg.speed = Speed;
    msg.angle = Angle;
    pub.publish(msg);
}

int main(int argc, char ** argv)
{
    //change_val(); // 1.8 이하의 값들을 1.8로!! (64m 넘어갈 때 0으로 반환하는 것과 차이를 두기 위해)
    ros::init(argc, argv, "miro_drive");
    ros::NodeHandle nh;
    ros::Subscriber sub = nh.subscribe("/scan", 1, callback);
    pub = nh.advertise<xycar_msgs::xycar_motor>("xycar_motor", 1);
    ros::Duration(3).sleep(); // ready to connect lidar
    
    while(ros::ok())
    {
        ros::spinOnce();
        
        cal_avg(avg_val);
        
        int front = 0;
        int front_left = 0;
        int front_right = 0;
        int left = 0;
        int right = 0;
        bool fflag = false;
        bool frflag = false;
        bool flflag = false;
        bool lflag = false;
        bool rflag = false;

        // 각 5구간을 나눠서 진행 (front, left, right)
        // front
        for(int i =361; i < 400; i++){
            cout << Dist[i] << " ";
        }
        cout <<"\n" << "---------------------------------------------------------------------------------------------" <<"\n";
        for (int degree = 484; degree <505; degree++)
        {
            if (Dist[degree] <= 0.3 && Dist[degree] != 0)
            {
                front += 1;
            }
            if (front > 3)
            {
                fflag = true;
                break;
            }
        }
        for (int degree = 0; degree < 21; degree++)
        {   
            
            if (Dist[degree] <= 0.45 && Dist[degree] != 0)
            {
                front += 1;
            }
            if (front > 3)
            {
                fflag = true;
                break;
            }
        }
        // front_left
        for (int degree = 400; degree < 484; degree++)
        {   
            if (Dist[degree] <= 0.45 && Dist[degree] != 0)
            {
                front_left += 1;
            }
            if (front_left > 3)
            {
                flflag = true;
                break;
            }
        }
        // front_right
        for (int degree = 21; degree < 105; degree++)
        {
            if (Dist[degree] <= 0.4 && Dist[degree] != 0)
            {
                front_right += 1;
            }
            if (front_right > 3)
            {
                frflag = true;
                break;
            }
        }
        // right
        for (int degree = 105; degree < 147; degree++)
        {
            if (Dist[degree] <= 0.4 && Dist[degree] != 0)
            {
                right += 1;
            }
            if (right > 3)
            {
                rflag = true;
                break;
            }
        }
        // left
        for (int degree = 358; degree < 400; degree++)
        {
            if (Dist[degree] <= 0.4 && Dist[degree] != 0)
            {
                left += 1;
            }
            if (left > 3)
            {
                lflag = true;
                break;
            }
        }
        if(fflag == false && frflag == false && flflag == false){
            drive_go(10, 0); 
            ROS_INFO("case: 1");
        }
        else if(fflag== true && frflag== false &&flflag ==false){
            int ret = -1;
            int d= 0;
            for(int i =0; i < 5; i++){
                if(avg_val[i] == 0){
                    d = i;
                    break;
                }
                if(ret <= avg_val[i]){
                    d = i;
                    ret = avg_val[i];
                }
                else{
                    ;
                }
            }
            drive_go(-10,-dic_ang[d]);
            ROS_INFO("case: 2");
            ros::Rate(1).sleep();
        }

        else if(fflag == false && frflag == true && flflag == false){
            drive_go(12, 50);
            ROS_INFO("case: 3");
        }
        else if(fflag == false && frflag == false && flflag == true){
            drive_go(12, -50);
            ROS_INFO("case: 4");
        }
        else if(fflag == true && frflag == true && flflag == false){
            drive_go(-10, -50);
            ros::Rate(1).sleep();
            ROS_INFO("case: 5");
        }
        
        else if(fflag == true && frflag == false && flflag == true){
            
            drive_go(-10, 50);
            ros::Rate(1).sleep();
            ROS_INFO("case: 6");
        }
        else if(fflag == true && frflag == true && flflag == true){
            
            if(rflag == false && lflag == true){
                drive_go(15, -50);
                ROS_INFO("case: 7");
            }
            else if(rflag == true && lflag == false){
                drive_go(15,50);
                ROS_INFO("case: 8");
            }
            else{
                ROS_INFO("case: nothing");
                int rret = -1;
                int dd= 0;
                for(int i =0; i < 5; i++){
                    if(avg_val[i] == 0){
                        dd = i;
                        break;
                    }
                    if(rret <= avg_val[i]){
                        dd = i;
                        rret = avg_val[i];
                    }
                    else{
                        ;
                    }
                }
                drive_go(-10,-dic_ang[dd]);
                ROS_INFO("case: 2");
                ros::Rate(1).sleep();
            }
            
        }
    }
    return 0;
}

File: src/miro_drive.py
#!/usr/bin/env python3
#-*- coding: utf-8 -*-
import rospy, time
from sensor_msgs.msg import LaserScan
from xycar_msgs.msg import xycar_motor

motor_msg = xycar_motor()
distance = []
pub = None

def callback(data):
    global distance, motor_msg
    distance = data.ranges

#motor_msg에 speed, angle 담아서 전송    
def drive_go(Speed, Angle):
    global motor_msg, pub
    motor_msg.speed = Speed
    motor_msg.angle = Angle
    pub.publish(motor_msg)
    

def main():
    global pub
    rospy.init_node('miro_drive')
    rospy.Subscriber('/scan', LaserScan, callback, queue_size = 1)
    pub = rospy.Publisher('xycar_motor', xycar_motor, queue_size = 1)
    time.sleep(3) #ready to connect lidar
    while not rospy.is_shutdown():
        
        front = 0
        front_left = 0
        front_right = 0
        left = 0
        right = 0
        fflag = False
        frflag = False
        flflag = False

        #각 5구간을 나눠서 진행(front, left. right)
        # front
        for degree in range(0,30):
            if distance[degree] <= 0.3 and distance[degree] != 0: #30cm 이내 진입시, 방해물이 64m 넘을 때 0값으로 들어와서 distance[degree] != 0로 오류 방지
                front += 1
            if front > 3: #해당 값 이내에 3번 진입시 front flag on
                fflag = True
                break
        # front_left
        for degree in range(-60,0):
            if distance[degree] <= 0.2 and distance[degree] != 0:
                front_left += 1
            if front_left > 3:
                flflag = True
                break
        #front_right
        for degree in range(30,90):
            if distance[degree] <= 0.2 and distance[degree] != 0:
                front_right += 1
            if front_right > 3:
                frflag = True
                break
        #right    
        for degree in range(90,120):
            if distance[degree] <= 0.3 and distance[degree] != 0: #30cm 이내 진입시, 방해물이 64m 넘을 때 0값으로 들어와서 distance[degree] != 0로 오류 방지
                right += 1
            if right > 3: #해당 값 이내에 3번 진입시 front flag on
                rflag = True
                break
        #left
        for degree in range(-90,-60):
            if distance[degree] <= 0.3 and distance[degree] != 0: #30cm 이내 진입시, 방해물이 64m 넘을 때 0값으로 들어와서 distance[degree] != 0로 오류 방지
                left += 1
            if left > 3: #해당 값 이내에 3번 진입시 front flag on
                lflag = True
                break

        #front_flag, left_flag, right_flag의 조합으로 8가지 경우로 나눔
        #case1 방해물 없는 경우 :직진
        if fflag == False and frflag == False and flflag == False:
            drive_go(10, 0)
        #case2 전면만 막힌경우: 우회전으로 후진
        elif fflag == True and frflag == False and flflag == False:
            drive_go(-10, 20)
        #오른쪽 방해물 있는 경우: 좌회전으로 전진    
        elif fflag == False and frflag == True and flflag == False:
            drive_go(10, -30)
        #왼쪽 방해물 있는 경우: 우회전으로 전진
        elif fflag == False and frflag == False and flflag == True:
            drive_go(10, 30)
        #전면과 오른쪽에 방해물 있는 경우: 좌회전으로 후진    
        elif fflag == True and frflag == True and flflag == False:
            drive_go(-10, -20)
        #전면과 왼쪽에 방해물 있는 경우: 우회전으로 후진
        elif fflag == True and frflag == False and flflag == True:
           drive_go(-10, 20)
        #전면, 왼쪽, 좌측, 방해물이 있는 경우: 직선으로 후진
        elif fflag == True and frflag == True and flflag == True:
            if rflag == False and lflag == True:
                drive_go(10, 50)
            elif rflag == True and lflag == False:
                drive_go(10,-50)
            else:
                drive_go(-10, 0)



if __name__ == '__main__':
    main()